New CSC segment builder - RoadUsage(RU)

nvoytish opened the pull request:
A new CSC segment building algorithm.
Name - RoadUsage(RU). Developers - V.Palichik and N. Voytishin
This algorithm shows better performance for all types of particles that can be registered by the CMS endcap muon system comparing to other existing algorithms. The best improvement is achieved for high Pt muons and events with high RecHit multiplicity in chamber.
The RoadUsage(RU) CSC segment building algorithm is added in the configuration files.
The file CSCSegmentAlgorithmRU_cfi.py contains the parameters set for the algorithm.
In order to use the RU algo, in line nr. 13 of the cscSegments_cfi.py file the 5th algo needs to be chosen.

nvoytish committed:
Config files with the 5th CSCSegBulder

The RoadUsage(RU) CSC segment building algorithm is added in the configuration files.
The file CSCSegmentAlgorithmRU_cfi.py contains the parameters set for the algorithm.
In order to use the RU algo, in line nr. 13 of the cscSegments_cfi.py file the 5th algo was chosen.

nvoytish committed:
RoadUsage(RU) CSC SegmentBuilder added

A new CSC segment building algorithm.
Name - RoadUsage(RU). Developers - V.Palichik and N. Voytishin
This algorithm shows better performance for all types of particles that can be registered by the CMS endcap muon system comparing to other existing algorithms. The best improvement is achieved for high Pt muons and events with high RecHit multiplicity in chamber.

nvoytish committed:
fixed blank lines problem

cmsbuild commented:
A new Pull Request was created by  @nvoytish  for CMSSW_8_1_X.

It involves the following packages:

RecoLocalMuon/CSCSegment


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@bellan, @abbiendi, @jhgoh this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a8f5f070f7e6b6e437d8da5af34f0fa78a0cde51
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13311/summary.html

slava77 commented:
@ptcox please check this and confirm it's OK from CSC DPG perspective
I would also like to suggest you get yourself added to watchers of CSC-related packages:
edit  https://github.com/cms-sw/cms-bot/blob/master/watchers.yaml online ans submit a PR or open an issue on https://github.com/cms-sw/cms-bot

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13311/summary.html

ptcox commented:
I did a quick scan but there's no way I can possibly give a sensible review within a few days. The basics look OK to me and it seems the config is now compatible with what we need for CSC segment builders to turn on or off specific algorithms. 
Once thing I notice is https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13311/llvm-analysis/report-2bd896.html#EndPath
which I don't understand. Does anyone?
On the other hand, CSCSegAlgoRU::stationNumber() is not well  named. It does not return a station number but a chamber type,. And its functionality can be replaced by the existing CSCDetId::iChamberType(), except that returns a value 1-10 so will need 1 subtracted to match stationNumber's 0-9.
Otherwise I am happy to have this algorithm released, so we can start more detailed tests in parallel with the standard algorithm CSCSegAlgoST. *We do NOT want CSCSegAlgoRU to yet become the default algorithm*.

slava77 commented:
about the static analyzer issues: I'm not sure it checks the range of values based on the implementation of the method ... however, the method itself can technically return "-1" unless one knows the assumptions/ranges of the stations.
You can add an assert(iStn >=0 && iStn <10) to help the static code analysis.

nvoytish committed:
switched back to ST algo as default

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish committed:
stationNumber() replaced by id.iChamberType()

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish committed:
stationNumber() function deleted

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish commented:
ST algo back as the default one. stationNumber() function is omitted and the standard iChamberType() is used instead. Please test.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: f6a67695d787c3a15be53d346e5c7f1810c19e60

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13358/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
50202.0 step2
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step2_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>4.22 step1
<pre>DAS Error</pre>134.911 step1
<pre>DAS Error</pre>140.53 step1
<pre>DAS Error</pre>1000.0 step1
<pre>DAS Error</pre>1001.0 step1
<pre>DAS Error</pre>1003.0 step1
<pre>DAS Error</pre>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

nvoytish commented:
That's strange. I run the tests as it is indicated at http://cms-sw.github.io/PRWorkflow
and get in the end 10 10 9 7 5 1 1 tests passed, 0 0 0 0 0 0 0 failed

ptcox commented:
I couldn't understand the log output. I hope Slava knows how to!
________________________________
From: nvoytish [notifications@github.com]
Sent: 03 June 2016 22:56
To: cms-sw/cmssw
Cc: Tim Cox; Mention
Subject: Re: [cms-sw/cmssw] New CSC segment builder - RoadUsage(RU) (#14718)


That's strange. I run the tests as it is indicated at http://cms-sw.github.io/PRWorkflow
and get in the end 10 10 9 7 5 1 1 tests passed, 0 0 0 0 0 0 0 failed

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/cms-sw/cmssw/pull/14718#issuecomment-223692386>, or mute the thread<https://github.com/notifications/unsubscribe/AE2FnlXMvp957QHUYbE_HHbDeeLoEBs8ks5qIJTsgaJpZM4IrSXh>.

cmsbuild commented:
+1
Tested at: f6a67695d787c3a15be53d346e5c7f1810c19e60
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13367/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13367/summary.html

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  : CSCSegmentAlgorithm(ps), myName("CSCSegAlgoRU"), sfit_(0) { 
+
+  doCollisions = ps.getParameter<bool>("doCollisions"); 
+  chi2_str_   = ps.getParameter<double>("chi2_str_"); 
```
was there a reason to have ``` chi2_str_ ``` instead of ``` chi2_str ``` ?
style oddities can lead to typos

nvoytish replied:
there is a chi2_str variable at line 706 used in baseline() function. 
Are you suggesting to change them? for example chi2_str_ -> chi2_str and chi2_str -> chi2_str_baseline.

slava77 replied:
My comment was only about the name of the configuration parameter "chi2_str_".
The c++ double/float representation of it can have a different name.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  } 
+
+  else if ( z1 < 0. ) { 
+    if ( z1 < z6 ) { 
```
the two if-if pairs above look like a single
``` if (std::abs(z1) > std::abs(z6) ) ``` 
this looks more readable and uses much less code

nvoytish replied:
You are right. This part was inherited from the SK algo and I didn't pay attention to this. I can change this too.

slava77 commented:
@nvoytish 
I'm slowly going through the code. I'll post a message when I'm done.
You can wait to the end or update as we go.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+
+    for(unsigned int n_seg_min = 6u; n_seg_min > 2u + iadd; --n_seg_min){ 
+      BoolContainer common_used(rechits.size(),false); 
+      std::array<BoolContainer, 120> common_used_it = {}; 
```
If I'm not mistaken 
``` common_used_it {} ``` will initialize to zeroes
```common_used_it = {}; ``` will leave elements of common_used_it uninitialized

slava77 replied:
doesn't matter, actually, since all 120 elements are initialized below.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  if(err_stpos_h1>0.25*strip_width[iStn] || err_stpos_h2>0.25*strip_width[iStn])dphi_incr = 0.5*strip_width[iStn]; 
+  float h1p = gp1.phi(); 
+  float h2p = gp2.phi(); 
+  float dphi12 = h1p - h2p; 
```
``` reco::deltaPhi(gp1,gp2) ``` is shorter

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+      BoolContainer common_used(rechits.size(),false); 
+      std::array<BoolContainer, 120> common_used_it = {}; 
+      for (unsigned int i = 0; i < common_used_it.size(); i++) { 
+	common_used_it[i] = common_used; 
```
the last time we exchanged igprof results, deque initialization and cleanup was taking 1/3 of the module time
https://slava77sk.web.cern.ch/slava77sk/reco/cgi-bin/igprof-navigator/nvoytish.ig.050516/44
Maybe a 2-D array with fixed size would work better (if we want to avoid vector bool): it will be size 150x120. The only disadvantage is having to keep track of the hits size separately.

nvoytish replied:
Correct me if I'm wrong. 
If you suggest something like `bool common_used_it[150][120]={{false}};` then I will have to change 
`  // Initialize flags that a given hit has been allocated to a segment 
  BoolContainer used(rechits.size(), false); 
  BoolContainer used3p(rechits.size(), false); `
to bool arrays also, because all three of them are used by function 
`void CSCSegAlgoRU::flagHitsAsUsed(const ChamberHitContainer& rechitsInChamber,  
				  BoolContainer& used ) const `

If that will help to shorten the time spent on this module, I will change it.

slava77 replied:
a test with that option will be needed, indeed.
Since it's a technical improvement, this doesn't have to delay this PR.

nvoytish replied:
Do you mean that I can change it later, or I need to do this in this PR?

slava77 replied:
It doesn't have to be in this PR.
It would be nice to have it though.

slava77 commented:
@nvoytish what's the combinatorial complexity of the algorithm?
Will we still be able to run it on beam splash events?
@ptcox maybe you can suggest a file to test in order to compare timing of cscSegments with the ST  and the RU algo.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  while(it != segments.end()) { 
+    if((*it).nRecHits() == 3){ 
+      bool found_common = false; 
+      std::vector<CSCRecHit2D> theseRH = (*it).specificRecHits(); 
```
specificRecHits() returns a reference.
I don't see any modifications to ``` theseRH ``` in this scope. 
Please change to ``` const std::vector<CSCRecHit2D>& theseRH = ```

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+	  int RH_wg = sh1->hitWire();
+	  std::vector<CSCRecHit2D>::const_iterator sh; 
+	  for(sh = theseRH.begin(); sh != theseRH.end(); ++sh){ 
+	    CSCDetId idRH = (CSCDetId)(*sh).cscDetId();
```
cscDetId is already of CSCDetId type.
Why not just ``` CSCDetId idRH = sh->cscDetId(); ```
?

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.h:
```diff
@@ -0,0 +1,160 @@
+    int minLayersApart;
+    bool debugInfo;
+
+    CSCSegFit* sfit_;
```
based on use pattern, it is better to change this to ``` std::unique_ptr<CSCSegFit> ```
after that all ``` delete sfit_``` can be removed and one wouldn't need to worry that frequent instantiations of this object leak memory.

nvoytish replied:
I am a bit confused. Using `std::unique_ptr<CSCSegFit>` in the header file, allows me to not care about cleaning `sfit_` in the entire source file? It is copied and reseted quite often.

slava77 replied:
yes, the smart pointer (unique_ptr) will handle the pointed object deletion

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  SVector6 se;  
+
+  ChamberHitContainer buffer1; 
+  buffer1.clear(); 
```
add ```buffer.reserve(init_size)```

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  ChamberHitContainer::iterator rh_to_be_deleted_1; 
+  ChamberHitContainer::iterator rh_to_be_deleted_2; 
+  if ( nhits > n_seg_min && (chi2_str/(nhits-2)) > chi2_str_*chi2D_iadd){ 
+    for (ChamberHitContainer::iterator i1 = proto_segment.begin(); i1 != proto_segment.end();++i1) { 
```
rh_to_be_deleted\* and the loop here and further below can all be const_iterator

nvoytish replied:
iterator changed to const_iterator in the entire baseline() function plus one identical change in replaceHit() function

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+  return (intercept + slope*layer); 
+} 
+
+float  CSCSegAlgoRU::fitX(SVector6 points, SVector6 errors, float &chi2_str){ 
```
this is the same as ``` fitX_ir(points, errors, -1, hi2_str) ```
do we need to full copied implementation?

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+} 
+
+float CSCSegAlgoRU::fit_sp(SVector6 points, SVector6 errors, int layer){ 
+
```
you could add iDelta (3.5 or 0  ) to fitX_ir and there would be no need for this large copy-pasted block inside fit_sp

nvoytish replied:
fit_sp() function used previously for validation of the code and not used anymore. To be deleted.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:
```diff
@@ -0,0 +1,1038 @@
+} 
+
+float CSCSegAlgoRU::fitX_ir(SVector6 points, SVector6 errors, int ir, float &chi2_str){ 
+
```
this method implementation looks equivalent to 
``` fitX_ir2(points, errors, ir, -1, chi2_str) ```
(30 lines less code)

nvoytish replied:
fitX, fitX_ir & fitX_ir2 changed to a single function fitX.

slava77 commented on RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.h:
```diff
@@ -0,0 +1,160 @@
+    // non-standard i.e. packed-bit implementation which is not a standard STL container. 
+    // We don't need what it offers and it could lead to unexpected trouble in the future.
+
+    typedef std::deque<bool> BoolContainer;
```
while ```vector<bool>``` doesn't have everything as functional as a regular type vector,
have you actually found an issue using it?
Based on the code in this PR, it looks like  ```vector<bool>```  will work

nvoytish replied:
this typedef was inherited from the SK algo. 
I will change it to ` vector<bool>`.

slava77 commented:
@nvoytish 
I'm done looking at the code.
There is a relatively small set of comments.
Please follow up.
Thank you.

nvoytish commented:
@slava77 Thanks for the comments. I am attending/organizing  а conference right now and unfortunatly have no time to work on any other tasks. I'll start changing the code on weekend.

slava77 commented:
ping
(a weekly check of status/progress)

nvoytish commented:
While analyzing Gif++ data we found some strange behavior of the new algo. Need a couple days to investigate(if that's ok with you).

nvoytish committed:
partial code optimization 

partial code optimization performed after slava77 suggestions.

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish committed:
chi2_str_ -> chi2_str

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish committed:
partial code optimization

some unused function deleted
deque changed to vector
function fitX() changed in order to be universal.

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
@nvoytish 
please let me know if you are done with the changes

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

cmsbuild commented:
-1

Tested at: 5a42720a7d69fc3cc1e1bd71354d813eb69df439

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13608/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling  /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/Utilities/StaticAnalyzers/src/EDMPluginDumper.cc 
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:12:0:
/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/Math/interface/deltaPhi.h: In instantiation of 'T reco::reduceRange(T) [with T = Geom::Phi<float>]':
/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/Math/interface/deltaPhi.h:47:23:   required from 'T reco::deltaPhi(T, T) [with T = Geom::Phi<float>]'
/build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:462:46:   required from here
/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/Math/interface/deltaPhi.h:16:34: error: the type 'const Geom::Phi<float>' of constexpr variable 'o2pi' is not literal
    constexpr T o2pi = 1./(2.*M_PI);
                                  ^
In file included from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/GeometryVector/interface/sseBasic2DVector.h:4:0,
                 from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/GeometryVector/interface/Basic2DVector.h:8,
                 from /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/GeometryVector/interface/PV2DBase.h:4,
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoST.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/src/SealModule.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/test/CSCSegmentReader.cc 
>> Compiling edm plugin /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/test/CSCSegmentVisualise.cc 
In file included from /build/cmsbuild/jenkins-workarea/workspace/ib-any-integration/CMSSW_8_1_X_2016-06-21-1100/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:12:
/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/Math/interface/deltaPhi.h:16:16: error: constexpr variable cannot have non-literal type 'const Geom::Phi<float>'
   constexpr T o2pi = 1./(2.*M_PI);
               ^
/cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-21-1100/src/DataFormats/Math/interface/deltaPhi.h:47:12: note: in instantiation of function template specialization 'reco::reduceRange<Geom::Phi<float> >' requested here
    return reduceRange(phi1 - phi2);
           ^
</pre>

slava77 commented:
I guess I should have waited for an answer to https://github.com/cms-sw/cmssw/pull/14718#issuecomment-227227659
clearly, this was not ready, since it didn't compile
It's more practical to push the commits from local to remote only after they are reasonably checked to be good.

In palces where deltaPhi is used and the underlying type is from globalPosition(), use barePhi() instead of phi()

nvoytish commented:
I was changing the code. The test were triggered too early. The phi issue will be fixed. Still working on replacing `CSCSegFit* sfit_` with `std::unique_ptr<CSCSegFit>`. There are some places in the code with oldfit witch are tricky for me.

nvoytish committed:
phi()->barePhi() changed

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish commented:
@slava77 change phi()->barePhi(). I still have problems with replacing `CSCSegFit* sfit_` with `std::unique_ptr<CSCSegFit>`. May be we can leave it for a future PR along with the BoolContainer changed to a simple bool array.
I run the tests with the current version of the code and got: 10 10 9 7 5 1 1 tests passed, 0 0 0 0 0 0 0 failed
The only thing that is strange to me is the following:
[nvoytish@lxplus059 src]$ git cms-checkdeps -a
>> Checking RecoLocalMuon/CSCSegment CMSSW_8_0_1
   x RecoLocalMuon/CSCSegment/python/cscSegments_cfi.py
   x RecoLocalMuon/CSCSegment/src/SealModule.cc
Checking out these packages: 1
RecoLocalMuon/Configuration (python)
error: Entry 'RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc' not uptodate. Cannot merge.

Is this ok?

slava77 commented:
``` I still have problems with replacing CSCSegFit* sfit_  ```
what is the problem?

nvoytish commented:
for example how to initialize or overwrite such objects:
` CSCSegFit* oldfit = new CSCSegFit( *sfit_ );`
or
`   sfit_ = new CSCSegFit( theChamber, proto_segment );
`
?

And also at the beginning I get an error that I don't quite get how to omit: 
`/afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.cc:21:61: error: call of overloaded 'unique_ptr(int)' is ambiguous
   : CSCSegmentAlgorithm(ps), myName("CSCSegAlgoRU"), sfit_(0) {                                                 
`
or `sfit_(0)` is not needed any more?

slava77 commented:
was this PR closed on purpose?

slava77 commented:
I see, you issues with unique_ptr should be cleared from a C++ documentation.

- for the initializer list in the constructor
null initialization is a default constructor, so, either ``` sfit_(nullptr) ``` or not needed anymore
- for the re-assignment ``` sfit_.reset(new CSCSegFit(  arguments ) ) ```

nvoytish commented:
I realy don't have a clue how that could happen.

nvoytish commented:
Actually now I know how this PR got closed  - my fault. Sorry about that.
I will fix the issue with unique_ptr, run tests and I will let you know when it's all done.

slava77 commented:
On 6/22/16 12:35 PM, nvoytish wrote:
> Actually now I know how this PR got closed - my fault. Sorry about that.

The buttons to close and to comment are next to each other.


> I will fix the issue with unique_ptr, run tests and I will let you know
> when it's all done.

Thank you.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14718#issuecomment-227852879>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbmM8jJT3gBmAHWm-4MaAkTwB9sTPks5qOY5tgaJpZM4IrSXh>.
>

nvoytish commented:
@slava77 I added unique_ptr to the code, but I am not quite sure that I did it right. There are still errors so I am not committing the code yet (or it will be more appropriate to commit, even if it's not done). Could you comment on the current state of the last function from the source file:
 void CSCSegAlgoRU::increaseProtoSegment(const CSCRecHit2D* h, int layer, int chi2_factor) {
                                                                                                                     
  // Creates a new fit    
  std::unique_ptr<CSCSegFit> oldfit;
  oldfit.reset(new CSCSegFit( theChamber, proto_segment ));
 bool ok = addHit(h, layer);
                                                                                                                      
  if ( !ok || ( (sfit_->ndof() > 0) && (sfit_->chi2()/sfit_->ndof() >= chi2Max)) ) {
     // reset to original fit                                                                                                                                                        
    sfit_ = std::move(oldfit);
  }
}

and one more thing. Everything compiles besides the last step SealModule.cc. The error is refering to the destructor in the header file ` virtual ~CSCSegAlgoRU() {};`

>> Compiling edm plugin /afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/SealModule.cc 
In file included from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/memory:81:0,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/internal/../tbb_stddef.h:421,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/internal/_concurrent_unordered_impl.h:30,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/concurrent_unordered_map.h:27,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/PluginManager/interface/PluginFactoryBase.h:27,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/PluginManager/interface/PluginFactory.h:26,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/Framework/src/Factory.h:4,
                 from /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/Framework/interface/MakerMacros.h:4,
                 from /afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/SealModule.cc:2:
/cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h: In instantiation of 'void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = CSCSegFit]':
/cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h:236:16:   required from 'std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = CSCSegFit; _Dp = std::default_delete<CSCSegFit>]'
/afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.h:68:29:   required from here
/cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h:74:22: error: invalid application of 'sizeof' to incomplete type 'CSCSegFit'
  static_assert(sizeof(_Tp)>0,
                      ^
gmake: *** [tmp/slc6_amd64_gcc493/src/RecoLocalMuon/CSCSegment/src/RecoLocalMuonCSCSegment/SealModule.o] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2


sorry if the questions are too stupid, but I consider you the expert in this field and it will be better to do it right from one try rather then redo it afterward.

slava77 commented:
On 6/23/16 5:43 AM, nvoytish wrote:
> @slava77 <https://github.com/slava77> I added unique_ptr to the code,
> but I am not quite sure that I did it right. There are still errors so I
> am not committing the code yet (or it will be more appropriate to
> commit, even if it's not done). Could you comment on the current state
> of the last function from the source file:
> void CSCSegAlgoRU::increaseProtoSegment(const CSCRecHit2D* h, int layer,
> int chi2_factor) {
> 
> // Creates a new fit
> 
> std::unique_ptr oldfit;
> oldfit.reset(new CSCSegFit( theChamber, proto_segment ));
> bool ok = addHit(h, layer);
> 
> if ( !ok || ( (sfit_->ndof() > 0) && (sfit_->chi2()/sfit_->ndof() >=
> chi2Max)) ) {
> // reset to original fit
> 
> sfit_ = std::move(oldfit);
> }
> }


this compiles, right?



> 
> and one more thing. Everything compiles besides the last step
> SealModule.cc. The error is refering to the destructor in the header
> file |virtual ~CSCSegAlgoRU() {};|
> 
>         Compiling edm plugin
>         /afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/SealModule.cc
> 
>         In file included from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/memory:81:0,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/internal/../tbb_stddef.h:421,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/internal/_concurrent_unordered_impl.h:30,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/tbb/44_20151115oss/include/tbb/concurrent_unordered_map.h:27,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/PluginManager/interface/PluginFactoryBase.h:27,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/PluginManager/interface/PluginFactory.h:26,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/Framework/src/Factory.h:4,
>         from
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/cms/cmssw/CMSSW_8_0_1/src/FWCore/Framework/interface/MakerMacros.h:4,
>         from
>         /afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/SealModule.cc:2:
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h:
>         In instantiation of 'void
>         std::default_delete<_Tp>::operator()(_Tp/) const [with _Tp =
>         CSCSegFit]':
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h:236:16:
>         required from 'std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with
>         _Tp = CSCSegFit; _Dp = std::default_delete]'
>         /afs/cern.ch/user/n/nvoytish/CMSSW_8_0_1/src/RecoLocalMuon/CSCSegment/src/CSCSegAlgoRU.h:68:29:
>         required from here
>         /cvmfs/cms.cern.ch/slc6_amd64_gcc493/external/gcc/4.9.3/include/c++/4.9.3/bits/unique_ptr.h:74:22:
>         error: invalid application of 'sizeof' to incomplete type
>         'CSCSegFit'


The compiler is telling you that it needs a header that defines
CSCSegFit included in the same file where the destructor is defined.
Your version in github now has CSCSegFit forward declared




>         static_assert(sizeof(_Tp)>0,
>         ^
>         gmake: */*
>         [tmp/slc6_amd64_gcc493/src/RecoLocalMuon/CSCSegment/src/RecoLocalMuonCSCSegment/SealModule.o]
>         Error 1
>         gmake: *** [There are compilation/build errors. Please see the
>         detail log above.] Error 2
> 
> sorry if the questions are too stupid, but I consider you the expert in
> this field and it will be better to do it right from one try rather then
> redo it afterward.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14718#issuecomment-228037982>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbrJxoS-2IatEKgxjjEBd2cY0RYUMks5qOn9VgaJpZM4IrSXh>.
>

nvoytish commented:
@slava77 

> this compiles, right?
>

yes it does. Is there something wrong&

> The compiler is telling you that it needs a header

now it compiled without errors.
Thank you for the help.
I'll run the tests and commit the changes if the tests are ok.

slava77 commented:
On 6/23/16 6:30 AM, nvoytish wrote:
> @slava77 <https://github.com/slava77>
> 
>     this compiles, right?
> 
> yes it does. Is there something wrong&


No problem, apparently.
It sounded as if you had a question about this part of the code.


> 
>     The compiler is telling you that it needs a header
> 
> now it compiled without errors.
> Thank you for the help.
> I'll run the tests and commit the changes if the tests are ok.


Great, thank you.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14718#issuecomment-228050124>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbuMpuetoLBvhOK-yDYMuMLpkmxSSks5qOoqCgaJpZM4IrSXh>.
>

nvoytish committed:
std::unique_ptr introduced

nvoytish committed:
std::unique_ptr<CSCSegFit> introduced

cmsbuild commented:
Pull request #14718 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

nvoytish commented:
@slava77 have you triggered the tests already? I was still commiting when the cmsbuild comment appeared. Hope it will work anyway.

nvoytish commented:
I am still a litle bit concerned about the deque-> vector change. There was a comment, that I deleted. I

slava77 commented:
On 6/26/16 2:27 PM, nvoytish wrote:
> @slava77 <https://github.com/slava77> have you triggered the tests
> already? I was still commiting when the cmsbuild comment appeared. Hope
> it will work anyway.

There were no tests triggered yet.
The bot is just informing that you have made a commit.
It would make much more sense to me if you push your commits to the
remote only once you are done
with code updates in your local repository and have a tested version.
There is no need to do it one by one (also, bigger commits of the same
kind are generally better than separate ones for individual files).

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14718#issuecomment-228598935>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbmLLE9pI_0TQdYicL5pX-gX8sYOVks5qPnAogaJpZM4IrSXh>.
>

nvoytish commented:
please test

slava77 commented:
On 6/27/16 12:43 PM, nvoytish wrote:
> please test
> 

not everyone can trigger tests via the bot.

As of your last commit 23 hours ago it looked like some more changes are
coming in the code.
Are you done with commits?


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14718#issuecomment-228712231>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbrjEZrb2IpAVTbGVTEGqNbvda4u1ks5qP6lpgaJpZM4IrSXh>.
>

nvoytish commented:
yes I am done. Sorry for not being clear. 
Could you please trigger the tests?

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 832a6eaec29c5bc949c7e01142a41fc2b4f5a8ad
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13689/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14718/13689/summary.html

There are some workflows for which there are errors in the baseline:
 5.1 step 2
 The results for the comparisons for these workflows could be incomplete 
This means most likely that the IB is having errors in the relvals.The error does NOT come from this pull request

slava77 commented:
Tested locally in 810pre8 with the algorithm enabled: results are consistent with the presentation in the RECO meeting on June 29
    - CPU use performance is OK: the last version is about a factor of 3 faster (on ttbar PU35)  compared to f6a6769; the last version is still about x2.5 slower than the default: 2.8 ms/evt vs 1.1 ms/evt on Zmm PU35 events which is acceptable.
        - The algorithm time doesn't run away based on one example of Run 268006, Event 234, LumiSection 67: current cscSegments takes 6 s/evt here while the new algo takes 8 s/evt.
    - the number of CSC segments is reduced, mainly from the short segments side (3 hits), while chi2 of the segments increased substantially
(plot from Z->mm with PU35)
![all_sign723vsorig_zmmpuwf25206p0c_cscdetidcscsegmentsownedrangemap_cscsegments__reco_obj_collection__data__degreesoffreedom](https://cloud.githubusercontent.com/assets/4676718/16587578/35c42b62-4280-11e6-802d-06df0ca3a107.png)
![all_sign723vsorig_zmmpuwf25206p0c_log10cscdetidcscsegmentsownedrangemap_cscsegments__reco_obj_collection__data__chi2](https://cloud.githubusercontent.com/assets/4676718/16587592/4a3db446-4280-11e6-8068-587cb405ffd4.png)
The rate of fake-like (low pt, low number of matches) muons has decreased
![all_sign723vsorig_zmmpuwf25206p0c_recomuons_muons__reco_obj_numberofmatches](https://cloud.githubusercontent.com/assets/4676718/16587658/8bb67552-4280-11e6-9f4d-02d2beb087c4.png)

Among higher level object plots, a moderate decrease in efficiency is seen in plots with quality requirement (I think this is expected because the algorithm apparently includes hits from deltas or other secondaries, which are not accounted in the quality definition)
Muon pt10 gun:
<img width="886" alt="wf1320_staupd_eff075_eta" src="https://cloud.githubusercontent.com/assets/4676718/16588189/e7b1ea74-4282-11e6-8d07-a55ab201df23.png">
Mu pt 1000 gun events
<img width="864" alt="wf1322_staupd_eff075_eta" src="https://cloud.githubusercontent.com/assets/4676718/16588197/f05087bc-4282-11e6-8b5b-43beab0d3013.png">

I think that to enable this algorithm in default workflows we may need to create a CSC segments collection without pointing constraint to retain effective tagging of cosmics or beam halo in collision events. However, it doesn't looks like there is a significant degradation in cosmics or beam halo reco in specialized samples (workflows 4.22 and 8.0)

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14718	832a6ea
- new CSC algorithm is introduced, it is not enabled in standard workflows
- jenkins tests pass and comparisons with baseline show no differences

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

slava77 commented:
@davidlange6 
Please comment if you expect to request any changes here.
Thanks.

nvoytish commented:
@slava77 could you answer the folowing questions or give me a link to the code that did the validation histograms that you posted 4 days ago?
I still don't understand the chi2 hist. How could it have negative values?
Also I would like to understand how the efficiencies are calculated in order to get the idea why the new algo is worse for some eta.

slava77 commented:
@nvoytish 
the chi2 plot that I posted is log10(chi2) as you could see in the header of the plot.
There are no negative values of chi2.

If you want to reproduce, you can follow https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Compare_outputs_FWLite_style
" Compare outputs FWLite style "

slava77 commented:
On 7/9/16 1:15 PM, nvoytish wrote:
> Also I would like to understand how the efficiencies are calculated in
> order to get the idea why the new algo is worse for some eta.

the numerator is here
https://github.com/cms-sw/cmssw/blob/0397259dd747cee94b68928f17976224c037057a/Validation/RecoMuon/plugins/MuonTrackValidator.cc#L196

and the denominator is at
https://github.com/cms-sw/cmssw/blob/0397259dd747cee94b68928f17976224c037057a/Validation/RecoMuon/plugins/MuonTrackValidator.cc#L78

nvoytish commented:
now I got it. I thought the log10 is refered to the vertical axis.
Thank you. I will take a look at the efficiencies and will try for sure to improve them.

slava77 commented:
IIRC, @abbiendi developed the quality-based muon efficiencies
In case you need help understanding the definitions.

nvoytish commented:
@slava77 Thanks again for the explanation and the help)

davidlange6 commented:
+1
a general comment: many of the loops in in this code are complex - some comments to describe the algorithm would be very beneficial to those that look at this code in the future.
